File: typescript/src/days/day02.ts
// Import NodeJS helpers
import fs from 'fs';

// Import types
import InputFile from '../types/InputFile';

// ----------------- Helpers ----------------- //

const wrongRPSScore = (game: string): number => {
  if (game === 'A X') return 4;
  if (game === 'A Y') return 8;
  if (game === 'A Z') return 3;
  if (game === 'B X') return 1;
  if (game === 'B Y') return 5;
  if (game === 'B Z') return 9;
  if (game === 'C X') return 7;
  if (game === 'C Y') return 2;
  if (game === 'C Z') return 6;
  return 0;
};

const rightRPSScore = (game: string): number => {
  if (game === 'A X') return 3;
  if (game === 'A Y') return 4;
  if (game === 'A Z') return 8;
  if (game === 'B X') return 1;
  if (game === 'B Y') return 5;
  if (game === 'B Z') return 9;
  if (game === 'C X') return 2;
  if (game === 'C Y') return 6;
  if (game === 'C Z') return 7;
  return 0;
};

// ------------- Main Functions -------------- //

/**
 * Day 2 Pt1 of Advent of Code 2022!
 * Scoring rock-paper-scissors games, the wrong way.
 * @author Benedikt Arnarsson
 */
const pt1 = async (filename: InputFile) => {
  const inputFile = await fs.promises.readFile(filename);

  const totalScore = inputFile
    .toString()
    .trim()
    .split('\n')
    .reduce((sum, game) => wrongRPSScore(game) + sum, 0);

  return totalScore;
};

/**
 * Day 2 Pt2 of Advent of Code 2022!
 * Scoring rock-paper-scissors games, the right way.
 * @author Benedikt Arnarsson
 */
const pt2 = async (filename: InputFile) => {
  const inputFile = await fs.promises.readFile(filename);

  const totalScore = inputFile
    .toString()
    .split('\n')
    .reduce((sum, game) => rightRPSScore(game) + sum, 0);

  return totalScore;
};

const day02 = [pt1, pt2];

export default day02;

File: elixir/lib/days/day02.ex
defmodule Days.Day02 do
  @moduledoc """
  Solutions to AoC 2022 Day 2: Rock Paper Scissors.
  """

  # Public Functions
  @spec pt1(binary()) :: integer()
  @spec pt2(binary()) :: integer()

  # Private Functions
  @spec wrong_rps_score(binary()) :: integer()
  @spec right_rps_score(binary()) :: integer()

  @doc """
  Part 1: Find rock-paper-scissors score (the wrong way).

  ## Examples

    iex> Days.Day02.pt1(day02_input) # filename
    <Solution to Day 2, part 1>

  """
  def pt1(file_name) do
    File.stream!(file_name)
      |> Stream.map(&String.trim/1)
      |> Enum.reduce(0, &(wrong_rps_score(&1) + &2))
  end

  @doc """
  Part 2: Find rock-paper-scissors score (the right way).

  ## Examples

    iex> Days.Day02.pt2(day02_input) # filename
    <Solution to Day 2, part 2>

  """
  def pt2(file_name) do
    File.stream!(file_name)
      |> Stream.map(&String.trim/1)
      |> Enum.reduce(0, &(right_rps_score(&1) + &2))
  end

  # --------------- Private Functions --------------- #

  defp wrong_rps_score(game) do
    case game do
      "A X" -> 4 # Rock, Rock
      "A Y" -> 8 # Rock, Paper
      "A Z" -> 3 # Rock, Scissors
      "B X" -> 1 # Paper, Rock
      "B Y" -> 5 # Paper, Paper
      "B Z" -> 9 # Paper, Scissors
      "C X" -> 7 # Scissors, Rock
      "C Y" -> 2 # Scissors, Paper
      "C Z" -> 6 # Scissors, Scissors
      _ -> raise "Malformed game: #{game}"
    end
  end

  defp right_rps_score(game) do
    case game do
      "A X" -> 3 # Rock, Rock
      "A Y" -> 4 # Rock, Paper
      "A Z" -> 8 # Rock, Scissors
      "B X" -> 1 # Paper, Rock
      "B Y" -> 5 # Paper, Paper
      "B Z" -> 9 # Paper, Scissors
      "C X" -> 2 # Scissors, Rock
      "C Y" -> 6 # Scissors, Paper
      "C Z" -> 7 # Scissors, Scissors
      _ -> raise "Malformed game: #{game}"
    end
  end
end
